Add docs for kNN search endpoint (#80378)
elastic/elasticsearch: docs/reference/search/knn-search.asciidoc

[[knn-search]]
=== kNN search API
++++
<titleabbrev>kNN search</titleabbrev>
++++

experimental::[]

Performs a k-nearest neighbor search and returns the matching documents.

[source,console]
----
PUT my-vector-index
{
  "mappings": {
    "properties": {
      "image_vector": {
        "type": "dense_vector",
        "dims": 3,
        "index": true,
        "similarity": "l2_norm"
      }
    }
  }
}

GET /my-vector-index/_knn_search
{
  "knn": {
    "field": "image_vector",
    "query_vector": [0.3, 0.1, 1.2],
    "k": 10,
    "num_candidates": 100
  },
  "_source": ["name", "date"]
}
----

> // TEST[setup:my_index]

Author: Thanks for catching these issues! I ended up hiding the mapping set-up.

> Is there a reason we went with _source instead of fields? Not a huge deal, but I figured fields would be preferred.

I initially had `fields` in some API examples, and one of the readers got very confused. Approaching the API for the first time, they thought `fields` referred to the vector fields for kNN. Unfortunately `fields` is not a widely-used parameter yet, so users would be learning two different concepts at the same time.

[[knn-search-api-request]]
==== {api-request-title}

`GET /<target>/_knn_search`

`POST /<target>/_knn_search`

[[knn-search-api-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have the `read`
<<privileges-list-indices,index privilege>> for the target data stream, index,
or alias.

[[knn-search-api-desc]]
==== {api-description-title}

The kNN search API performs a k-nearest neighbor search on a
<<dense-vector,`dense_vector`>> field. Given a query vector, it finds the _k_
closest vectors and returns those documents as search hits. To be used in the
kNN search API, the `dense_vector` field must

> <<index-vectors-knn-search, enable indexing>>.

Reviewer: Not a huge deal, but I'd consider moving the last sentence into the `knn.field` param definition. It's sort of a requirement for that param.

I overlooked this when we created the dense vector docs, but it may be good to note there that the `index` parameter isn't dynamic/updateable.

{es} uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to support
efficient kNN search. Like most kNN algorithms, HNSW is an approximate method
that sacrifices result accuracy for improved speed. This means the documents
returned are not always the true _k_ closest neighbors.

[[knn-search-api-path-params]]
==== {api-path-parms-title}

`<target>`::
(Optional, string) Comma-separated list of data streams, indices, and aliases
to search. Supports wildcards (`*`). To search all data streams and indices,

> use `*` or `_all`.
> 
> NOTE: The kNN search API does not support remote indices (as in cross-cluster
> search).

Reviewer: ```suggestion
use `*` or `_all`.
+
NOTE: The kNN search API doesn't support {ccs}.
```

Author: This is much simpler! I hadn't realized we used "CCS" in the context of APIs besides `_search` (field caps, now kNN, etc.)

[role="child_attributes"]
[[knn-search-api-query-params]]
==== {api-query-parms-title}

include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=routing]

[role="child_attributes"]
[[knn-search-api-request-body]]
==== {api-request-body-title}

`knn`::
(Required, object) Defines the kNN query to run.
+
.Properties of `knn` object
[%collapsible%open]
====
`field`::

> (Required, string) The name of the vector field to search against.

Reviewer: See above comment about noting the `dense_vector` and indexing requirement here.

`query_vector`::

> (Required, array of floats) The query vector.

Reviewer: Should we note this must be a `dense_vector` value?
```suggestion
(Required, array of floats) Query vector.
```

Reviewer 2: May be we can add that `query_vector` must have the same dims as an indexed field it searches against. Although it is also kind of obvious, so not sure if it is worth to add.

Author: I think it's clear it needs to be a dense vector value, since those are just arrays of floats?

Taking @mayya-sharipova's suggestion, I've added a note that it must have the same number of dimensions as the vector fields. I think it's good practice to document all the requirements, even if they're pretty self-evident.

`k`::

> (Required, integer) The final number of nearest neighbors to return as top
> hits. This value must be less than `num_candidates`.

Reviewer: ```suggestion
(Required, integer) Number of nearest neighbors to return as top hits. Must be
less than `num_candidates`.
```

`num_candidates`::

> (Required, integer) The number of nearest neighbor candidates to consider per
> shard. Increasing `num_candidates` tends to improve the accuracy of the final
> `k` results. This value cannot exceed 10,000.

Reviewer: We don't really describe the high-level process for gathering candidates from each shard and then collating the results. IIUC that and HSNW are the two main factors that can affect accuracy.

Should we save that for the high-level guide or mention in the description?

```suggestion
(Required, integer) Number of nearest neighbor candidates to consider per shard.
Cannot exceed `10000`. A higher value improves the accuracy of the results at
the cost of a slower search speed.
```

Reviewer 2: I agree with @jrodewig it would be nice to add more details here how this `num_candidates` works, for example: "{es}  collects from each shard  the top `num_candidates` results, and then merges collected from the shards results to get the top k results.  Increasing `num_candidates` tends to improve the accuracy of the final `k` results..."

Author: This is a nice idea I added the sentence " {es} collects `num_candidates` results from each shard, then merges
them to find the top `k` results". In the high level guide, I think we can step through how the kNN search is executed in more detail.

====

include::{es-repo-dir}/search/search.asciidoc[tag=fields-param-def]
include::{es-repo-dir}/search/search.asciidoc[tag=docvalue-fields-def]
include::{es-repo-dir}/search/search.asciidoc[tag=stored-fields-def]

> include::{es-repo-dir}/search/search.asciidoc[tag=source-filtering-def]

Reviewer: We typically try to order parameters alphabetically, but I kinda like this sort order.

Author: I can reorder these to be consistent. I was trying to emphasize that `fields` is what you usually want, so people don't accidentally start using `docvalue_fields`. But there's probably better ways to do this, like through examples and "how to" guides.

Author: Oh I just realized that if they were truly alphabetical, then `fields` and `docvalue_fields` would float above `knn`! I think that would really be unfortunate, as the `knn` section contains information that's important to understanding the API.

[role="child_attributes"]
[[knn-search-api-response-body]]
==== {api-response-body-title}

> A kNN search response has the exact same structure as a
> <<search-api-response-body, search response>>. However, certain sections take

Reviewer: I'd specifically mention search API here as we have a number of search-related APIs now.
```suggestion
A kNN search response has the same structure as a
<<search-api-response-body, search API response>>. However, certain sections take
```

Author: 👍  makes sense.

> <<search-api-response-body, search response>>. However, certain sections take
> on a meaning specific to kNN search:

Reviewer: ```suggestion
<<search-api-response-body, search response>>. However, certain sections have
a meaning specific to kNN search:
```

> * The <<search-api-response-body-score,document `_score`>> is determined by

Reviewer: Tried to reword this to avoid passive voice. Not a big deal if left as is tho.
```suggestion
* The <<search-api-response-body-score,document `_score`>> measures
```

Author: This is a subtle point where I think "is determined by" is more accurate/ clear -- the `_score` calculation is not that meaningful. It's a transformed version of the similarity that makes sure to obey Elasticsearch's score requirements.

the similarity between the query and document vector. See

> <<dense-vector-similarity, `similarity`>> for details.

Reviewer: ```suggestion
<<dense-vector-similarity, `similarity`>>.
```

* The `hits.total` object contains the total number of nearest neighbor

> candidates considered, which is `num_candidates * num_shards`. The `relation`

Reviewer: ```suggestion
candidates considered, which is `num_candidates * num_shards`. `hits.total.relation`
```

  },
  "_source": ["name", "date"]
}
----
// TEST[continued]

[[knn-search-api-request]]
==== {api-request-title}

`GET <target>/_knn_search`

`POST <target>/_knn_search`

[[knn-search-api-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have the `read`
<<privileges-list-indices,index privilege>> for the target data stream, index,
or alias.

[[knn-search-api-desc]]
==== {api-description-title}

The kNN search API performs a k-nearest neighbor (kNN) search on a
<<dense-vector,`dense_vector`>> field. Given a query vector, it finds the _k_
closest vectors and returns those documents as search hits.

{es} uses the https://arxiv.org/abs/1603.09320[HNSW algorithm] to support
efficient kNN search. Like most kNN algorithms, HNSW is an approximate method
that sacrifices result accuracy for improved search speed. This means the
results returned are not always the true _k_ closest neighbors.

[[knn-search-api-path-params]]
==== {api-path-parms-title}

`<target>`::
(Optional, string) Comma-separated list of data streams, indices, and aliases
to search. Supports wildcards (`*`). To search all data streams and indices,
use `*` or `_all`.

> NOTE: The kNN search API does not support {ccs}.

Reviewer 3: Is it something that we forbid explicitly ? I didn't test but it should work transparently since we use the search action internally.

Author: It indeed should already work transparently. I wrote this because I didn't want to set a precedent that it supports CCS while we're still figuring out the execution strategy -- especially around how to combine kNN results with term-based results. However it's not something we forbid explicitly (given our strategy of translating at the REST layer, it's not very simple to do). What are your thoughts?

Reviewer 3: IMO we're protected by the experimental status. CCS works out of the box and there's nothing that prevents us from removing the support later on. I don't see why we wouldn't though so I am not sure we need the NOTE at the moment.

Author: That reasoning makes sense to me, I'll remove the note to keep things simple.

+
.Properties of `knn` object
[%collapsible%open]
====
`field`::
(Required, string) The name of the vector field to search against.

`query_vector`::
(Required, array of float) The query vector.

`k`::
(Required, integer) The final number of nearest neighbors to return as top
hits. This value must be less than `num_candidates`.

`num_candidates`::
(Required, integer) The number of nearest neighbor candidates to consider per
shard. Increasing `num_candidates` tends to improve the accuracy of the final
`k` results. This value cannot exceed 10,000.
====

include::{es-repo-dir}/search/search.asciidoc[tag=fields-param-def]
include::{es-repo-dir}/search/search.asciidoc[tag=docvalue-fields-def]
include::{es-repo-dir}/search/search.asciidoc[tag=stored-fields-def]
include::{es-repo-dir}/search/search.asciidoc[tag=source-filtering-def]

[role="child_attributes"]
[[knn-search-api-response-body]]
==== {api-response-body-title}

A kNN search response has the exact same structure as a
<<search-api-response-body, search response>>. The
<<knn-search-api-response-body-score,`_score`>> of each hit is determined by
the similarity between the query and document vectors. See

> <<dense-vector-similarity, `similarity`>> for details.

Reviewer: I think this short paragraph is sufficient. We do something similar in the [EQL async search API docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/get-async-eql-search-api.html#get-async-eql-search-api-response-body). I wouldn't go through the hassle of tagging and including each field.

We may want to include a short bulleted list detailing any differences. For example, `hits` means something different for kNN.

Author: This is a good idea, I'll just add a section highlighting the differences.
